packages/reactivity/src/effect.ts

```diff
@@ -143,12 +143,23 @@ function cleanupEffect(effect: ReactiveEffect) {
   const { deps } = effect
   if (deps.length) {
     for (let i = 0; i < deps.length; i++) {
-      deps[i].delete(effect)
+      removeEffectFromDep(deps[i], effect)
     }
     deps.length = 0
   }
 }
 
+export function removeEffectFromDep(dep: Dep, effect: ReactiveEffect) {
+  dep.delete(effect)
+  if (dep.target && dep.size === 0) {
+    const depsMap = targetMap.get(dep.target)
+    if (depsMap) {
+      depsMap.delete(dep.key)
+    }
+    dep.target = dep.key = null
+  }
+}
+
```

Comment:
I considered using a `setTimeout()` here to defer the purging of the empty dependencies. Removing them immediately like this isn't strictly necessary, and could lead to some extra churn if empty dependencies are removed and then immediately re-added.

However, in the most common cases, the `trackOpBit` stuff avoids this extra churn, so I decided it wasn't worth the extra complexity.